File: 백준/Silver/1406. 에디터/에디터.swift
//
//  main.swift
//  BJ_1406
//
//  Created by 안홍범 on 3/7/25.
//

import Foundation

var words = readLine()!
let N = Int(readLine()!)!

func solution(_ words: String) -> String {
    var left = Array(words)
    var right = [Character]()
    
    for _ in 0..<N {
        let command = readLine()!
        switch command {
            case "L":
                if !left.isEmpty {
                    right.append(left.removeLast())
                }
            case "D":
                if !right.isEmpty {
                    left.append(right.removeLast())
                }
            case "B":
                if !left.isEmpty {
                    left.removeLast()
                }
            default:
                left.append(command.last!)
        }
    }
    
    return String(left + right.reversed())
}
    
print(solution(words))

File: 백준/Silver/1406. 에디터/에디터.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String words= bf.readLine();
        int N = Integer.parseInt(bf.readLine());
        String[][] commands = new String[N][2];

        for(int i = 0; i < N; i++){
            commands[i] = bf.readLine().split(" ");
        }

        System.out.printf("%s", solution(words, commands));
    }

    static String solution(String words, String[][] commands) {
        Character[] word = new Character[words.length()];
        for (int i = 0; i < words.length(); i++) {
            word[i] = words.charAt(i);
        }

        Stack<Character> left = new Stack<>();
        Stack<Character> right = new Stack<>();

        for (Character c : word) {
            left.push(c);
        }

        for (String[] command : commands) {
            switch (command[0]) {
                case "L":
                    if (!left.isEmpty()) {
                        right.push(left.pop());
                    }
                    break;
                case "D":
                    if (!right.isEmpty()) {
                        left.push(right.pop());
                    }
                    break;
                case "B":
                    if (!left.isEmpty()) {
                        left.pop();
                    }
                    break;
                case "P":
                    left.push(command[1].charAt(0));
                    break;
            }
        }

        return left.toString()
                .replaceAll("[\\[\\], ]", "")
                + new StringBuilder(right.toString())
                    .reverse()
                    .toString()
                    .replaceAll("[\\[\\], ]", "");
    }
}
